python/dgl/graphbolt/negative_sampler.py

```diff
@@ -0,0 +1,128 @@
+"""Negative samplers."""
+
+from _collections_abc import Mapping
+
+import torch
+from torchdata.datapipes.iter import Mapper
+
+
+class NegativeSampler(Mapper):
+    """
+    A negative sampler used to generate negative samples and return
+    a mix of positive and negative samples.
+    """
+
+    def __init__(
+        self,
+        datapipe,
+        negative_ratio,
+    ):
+        """
+        Initlization for a negative sampler.
+
+        Parameters
+        ----------
+        datapipe : DataPipe
+            The datapipe.
+        negative_ratio : int
+            The proportion of negative samples to positive samples.
+        """
+        super().__init__(datapipe, self._sample)
+        assert negative_ratio > 0, "Negative_ratio should be positive Integer."
+        self.negative_ratio = negative_ratio
+
+    def _sample(self, node_pairs):
+        """
+        Generate a mix of positive and negative samples.
+
+        Parameters
+        ----------
+        node_pairs : Tuple[Tensor] or Dict[etype, Tuple[Tensor]]
+            A tuple of tensors or a dictionary represents source-destination
+            node pairs of positive edges, where positive means the edge must
+            exist in the graph.
+
+        Returns
+        -------
+        Tuple[Tensor] or Dict[etype, Tuple[Tensor]]
+            A collection of edges or a dictionary that maps etypes to edges,
+            which includes both positive and negative samples.
+        """
+        if isinstance(node_pairs, Mapping):
+            return {
+                etype: self._generate(pos_pairs, etype)
+                for etype, pos_pairs in node_pairs.items()
+            }
+        else:
+            return self._generate(node_pairs, None)
+
+    def _generate(self, node_pairs, etype=None):
+        """Generate a mix of positive and negative samples for a given etype.
+
+        Parameters
+        ----------
+        node_pairs : Tuple[Tensor]
+            A tuple of tensors or a dictionary represents source-destination
+            node pairs of positive edges, where positive means the edge must
+            exist in the graph.
+        etype : (str, str, str)
+            Canonical edge type.
+        Returns
+        -------
+        Tuple[Tensor]
+            A mixed collection of positive and negative node pairs.
+        """
+        raise NotImplementedError
+
+    def _generate_negative_pairs(self, node_pairs, etype=None):
```

Comment:
def _post_process():
    """Covert negative pairs to a specified format.

Reply:
It has conflict with above comment which suggest using name `_sample_negative_pairs`. If so, I prefer `_post_process` as it is more accurate.